joomla/joomla-cms
# Joomla! Update: reinstall Joomla! and Update & Install features | seccond try

Pull Request for Issue #9469.
#### Summary of Changes

Added features to Joomla! Update:
- **Reinstall.** If you're already on the latest version you are given the option to re-download the full update package and install it on top of your site. Useful if someone has made core hacks and you want to undo them.
- **Upload & Update.** Now that we have removed the ability to install Joomla! as an extension to itself the users with no FTP access and/or no outbound server Internet access (e.g. behind a corporate DMZ firewall) are unable to update. We now have the option to upload the update package. Since this is dangerous and could potentially be abused you need to re-enter your login information after the upload and before the uploaded package is installed. Failure to comply removes the uploaded file.
#### Testing Instructions
###### First test
- install 3.4.8 from https://github.com/joomla/joomla-cms/releases/tag/3.4.8
- go to your database and run this command (it changes the update server of the update component to my own hosted version)

```
UPDATE  `#__update_sites` SET  `location` =  'http://www.jah-tz.de/downloads/extensions/dev/com_joomlaupdate.xml' WHERE  `location` = 'http://update.joomla.org/core/extensions/com_joomlaupdate.xml';
```
- go to Extensions -> Update
- install the com_joomlaupdate Update
- Go to com_joomlaupdate
- notice the layout of the new Update
- switch to the Tab "Upload & Install"
- download: https://github.com/joomla/joomla-cms/releases/download/3.5.0/Joomla_3.5.0-Stable-Update_Package.zip
- Upload the 3.5.0 update package.
- watch the progress bar
- see that you updated to 3.5.0 stable now.
###### Note

Please notice that as the 3.5.0 stable package don't include the changes here the new overview is gone for now. But if you follow the seccond test it come back :smile: 
###### seccond test
- install 3.5.0 stable or staging (you can also use the result of the first test)
- install com_patchtester
- apply this fix of com_patchtester: joomla-extensions/patchtester#133
- apply this patch
- go to com_joomlaupdate
- notice the new layout
- try the reinstall feature
- try the upload and install freature
#### For the code reviewer

The changes between the code by @nikosdion and me can be found here: https://github.com/nikosdion/joomla-cms/compare/feature/joomlaupdate-reinstall...zero-24:feature/joomlaupdate-reinstall
#### Some other things

The needed update xml for com_joomlaupdate can be found here: https://github.com/joomla/update.joomla.org/pull/6

The com_joomlaupdate package can be found here (it is the same as on my server for updating):
http://www.jah-tz.de/downloads/download.php?file=com_joomlaupdate.zip&amp;folder=extensions/dev/
#### Limitations of using the com_joomlaupdate standalone. (edit)

You can't use the reinstall feature as the standalone package don't update the updater lib file please see: https://github.com/joomla/joomla-cms/pull/9612#issuecomment-202142160
#### Thanks

Thanks goes to @nikosdion @wilsonge and @mbabker and com_patchtester (as there we can find how we should create a update packages)
